### PR TITLE
Add UI Support for up to 8x Internal Resolutions

### PR DESCRIPTION
**Pull request by RisingFog**
Adds new options from 5x up to 8x for Internal Resolutions.

### CODE REVIEW

**Comment by Alegend45** [comment]
Who the hell has monitors that big?

**Comment by phire** [comment]
Higher IRs also act as anti-aliasing.

**Comment by pauldacheez** [comment]
Not with the current downscaling algorithm, they don't. https://code.google.com/p/dolphin-emu/issues/detail?id=7746

**Comment by DZCreeper** [comment]
6x IR matches the width of a 4K display, so people with high end machines will see some use of this.

**Comment by degasus** [comment]
Does the "increase IR" hotkey still works fine?

**Comment by Linktothepast** [comment]
I will say just one thing, i am pretty certain there will be a lot of posts in the forums about slow speed with the emulator while having 8x IR... using their laptop... There are a lot of people that just use the highest value they can find and expect everything to run smooth afterwards.

**Comment by Pringo** [comment]
Thanks for adding this. imo this should satisfy Issue 7711 but I don't know if others would agree with that.

**Comment by neobrain** [comment]
This is getting quite ridiculous.

Other than that, I'd suggest just merging those resolutions into a "custom" item, and have the value configurable in discrete steps as a wxSlider.

**Comment by JMC47** [comment]
To everyone:  For screenshots, videos, and other features in Dolphin; having this available would be nice.

**Comment by adamd10d** [comment]
The only problem I see with this is accidentally clicking a crazy high IR
that your GPU can't handle, causing your machine to come grinding to a
halt, though that's probably just me being careless.

**Comment by Linktothepast** [comment]
Could there be suggestions depending on the monitor apart from just displaying the resolution? Also make auto integral the suggested value used.
Example:
"Auto (Window Size)"
"Auto (Multiple of 640x528) Default"
"1x Native (640x528)"
"1.5x Native (960x792)"
"2x Native (1280x1056) for 720p"
"2.5x Native (1600x1320)" 
"3x Native (1920x1584) for 1080p"
"4x Native (2560x2112) for WQHD"
"5x Native (3200x2640)"
"6x Native (3840x3168) for 4K UHD"
"7x Native (4480x3696)"
"8x Native (5120x4224)"

**Comment by bb010g** [comment]
:+1: for descriptions on the named resolutions.

**Comment by Rukario** [comment]
I don't like the idea to make anything beside 1x Native as default.

**Comment by JMC47** [comment]
I think auto is already default.  Also + 1 for naming them.  Maybe adding a description that higher IRs can be extremely slow on weaker (or even strong) hardware.

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 1]
```diff
@@ -357,16 +357,17 @@ VideoConfigDiag::VideoConfigDiag(wxWindow* parent, const std::string &title, con
 	// Internal resolution
 	{
 	const wxString efbscale_choices[] = { _("Auto (Window Size)"), _("Auto (Multiple of 640x528)"),
-		_("1x Native (640x528)"), _("1.5x Native (960x792)"), _("2x Native (1280x1056)"),
-		_("2.5x Native (1600x1320)"), _("3x Native (1920x1584)"), _("4x Native (2560x2112)"), _("Custom") };
+		_("1x Native (640x528)"), _("1.5x Native (960x792)"), _("2x Native (1280x1056) for 720p"), _("2.5x Native (1600x1320)"),
+		_("3x Native (1920x1584) for 1080p"), _("4x Native (2560x2112) for 1440p"), _("5x Native (3200x2640)"),
+		_("6x Native (3840x3168) for 4K"), _("7x Native (4480x3696)"), _("8x Native (5120x4224)"), _("Custom") };
```
Comment (Buddybenj) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: "8x Native (5120x4224) for 5K"

Also can't "1x Native (640x528)" just be "Native (640x528)"?

**Comment by mimimi085181** [comment]
Can you also update the texts that display when changing settings, where it says which IR is currently used, which efb type, fog etc. :
https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoCommon/RenderBase.cpp#L336

And pretty please with support for custom resolutions, and not just the ones you add to the UI?

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild (test, please ignore)

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild (test, please ignore)

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild (test, please ignore)

**Comment by RisingFog** [comment]
bump

**Comment by Pringo** [comment]
Partially a question and partially a suggestion but what exactly is the purpose of the 1.5x and 2.5x settings? I was under the impression that integer scaling was preferred for higher IRs. Would there be any harm in removing those to help clean up the list?

**Comment by bb010g** [comment]
@Pringo 1.5x or 2.5x is better than 1x or 2x for those with worse computers. Yes, it's not optimal, but it's better.

**Comment by Pringo** [comment]
I might be alone on this but if that's the only reason I'd support removing those options. 1.5x and 2.5x aren't going to provide major benefits over 1x and 2x and thanks to all the optimization work that's been done those users can probably get fullspeed now with 2x or 3x compared to the 4.0.2 release. I think removing these half step options would help clean up the interface.

**Comment by ADormant** [comment]
Related bug https://forums.dolphin-emu.org/Thread-custom-ir-issue

**Comment by RisingFog** [comment]
@ADormant not really in the scope of the PR tbh

**Comment by karasuhebi** [comment]
@ADormant create a new issue for it. I've been meaning to do it (as I have the same issue) but I hadn't seen confirmation from someone else experiencing it until now. :D

**Comment by ADormant** [comment]
Is anything still preventing this pull from being merged?

**Comment by ADormant** [comment]
Is there anything that is still preventing this pull request from being merged :question: It would be a convenient option.

**Comment by lioncash** [comment]
@ADormant You posted something similar like this yesterday. You don't need to keep bumping it; people likely saw the first message.

**Comment by RisingFog** [comment]
bump

**Comment by Stevoisiak** [comment]
LGTM

**Comment by kkkllleee** [comment]
@Linktothepast  Maybe we could just add a warning after we surpass the average resolution.

**Comment by JosJuice** [comment]
@kkkllleee: What do you mean by the average resolution? And in what way should such a warning be displayed?

**Comment by Buddybenj** [comment]
If the IR is higher than their monitor resolution, display an OSD message saying that they're wasting GPU power and it may cause unnecessary slowdowns.

**Comment by kkkllleee** [comment]
@JosJuice Nevermind A better alternative would be the one by @Buddybenj , where we make a check, comparing their IR against their MR, and display a warning if the internal resolution is greater.

**Comment by karasuhebi** [comment]
@Buddybenj Why is it wasting GPU power? I know Dolphin's downsampling is not the best but surely it helps somewhat with aliasing, no?

**Comment by DonelBueno** [comment]
@karasuhebi It helps more with sharpness than with aliasing.

**Comment by Buddybenj** [comment]
@karasuhebi [Issue 7746](https://code.google.com/p/dolphin-emu/issues/detail?id=7746)
This is mainly just to inform people with 1366x768 laptops that they're not going to see any difference if they use 8x native over 2.5x native. Believe it or not, lots of noobs just select the highest resolution possible.

**Comment by karasuhebi** [comment]
@DonelBueno @Buddybenj Alright so it helps with sharpness rather than aliasing, fair enough. Still, that doesn't mean the GPU power is wasted is it? I understand the whole situation with someone on a 1366x768 laptop setting IR to 8x native but I don't think Dolphin should we made to say it's waste of power. I guess it kinda is in his situation, but not across the board. Maybe just say that cranking up the IR past a certain point is starting to get into the realm of diminishing returns?

**Comment by karasuhebi** [comment]
@RisingFog thanks for implementing @mimimi085181's suggestion into RenderBase.cpp, but would you mind adding in support for more resolutions? I second her request on that. It seems like you have to make a case for each extra resolution which kinda sucks but I myself use higher resolutions at times and I'd like to be able to see them listed on that on-screen text.

For future reference, here's a month-old screenshot of the issue:
http://i.imgur.com/0IxDZby.png

**Comment by RisingFog** [comment]
@karasuhebi would you want them to be the full render resolution instead of 1x, 2x, etc?

**Comment by karasuhebi** [comment]
Multipliers is fine with me. Up to you. Thanks! :D

**Comment by RisingFog** [comment]
@karasuhebi in that case, what's in there should already handle any infinite IR multiplier

**Comment by karasuhebi** [comment]
Oh OK sweet. Sorry I don't really know much about programming. Must have misunderstood what was going on in there.

Thanks again!

**Comment by RisingFog** [comment]
bump

**Comment by ADormant** [comment]
Is there anything still precluding it from being merged? @RisingFog

**Comment by RisingFog** [comment]
Not to my knowledge, at least.

**Comment by neobrain** [comment]
There is an issue with adding more IR options, namely that people with low screen resolutions will not be able to select all of the available options anymore soon. As a solution, I suggest making the number of shown IR options dependent on the current Desktop resolution: That way, people don't end up with unselectable options, and also people with high-resolution setups can use as large IRs as they like, potentially up to 40xIR.

**Comment by RisingFog** [comment]
@neobrain limiting the selection based off of monitor resolution could be done. However, I feel that allowing selectable IR's above 8x is pointless due to current monitor sizes.

**Comment by ADormant** [comment]
@RisingFog  I agree with @neobrain that limiting resolutions based on user's monitor resolution would be the best option. Monitor resolutions increase all the time.

**Comment by Rukario** [comment]
Well, can't we just leave as is? It'd be user's own fault for picking the unsupported resolutions.

**Comment by neobrain** [comment]
... for reference, I was just kidding, referring back to my initial comment in this thread. It's a terrible idea, really.
Sorry for the confusion.

**Comment by bb010g** [comment]
Can we at least put in a custom IR option w/ an aspect ratio lock? That would probably be the cleanest way to deal with large IRs whilst still allowing for fullscreen with upscaling.

**Comment by shuffle2** [comment]
I agree with @neobrain

**Comment by sigmabeta** [comment]
Adding one more +1 to this PR. I was messing around with triple monitor Nvidia surround with 3x 1440p monitors. The 4x IR option wasn't enough to capably fill all of that space, but with the auto option, my GTX 970 was unable to maintain consistently full speed. 

8x would still not be enough for the three monitors, but it would probably be good enough.

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 2]
```diff
@@ -355,8 +359,22 @@ void Renderer::DrawDebugText()
...
+			res_text = "6x";
+			break;
+		case SCALE_7X:
+			res_text = "7x";
+			break;
+		case SCALE_8X:
+			res_text = "8x";
+			break;
+		default:
+			res_text = (g_ActiveConfig.iEFBScale - 3) + "x";
```
Comment (degasus) [inline Source/Core/VideoCommon/RenderBase.cpp]: Isn't this formulation valid for 3x IR and higher? So all of them can just fall back to default.

**Comment by Buddybenj** [comment]
Bump.

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 3]
```diff
@@ -355,16 +355,17 @@ VideoConfigDiag::VideoConfigDiag(wxWindow* parent, const std::string &title, con
 	// Internal resolution
 	{
 	const wxString efbscale_choices[] = { _("Auto (Window Size)"), _("Auto (Multiple of 640x528)"),
-		_("1x Native (640x528)"), _("1.5x Native (960x792)"), _("2x Native (1280x1056)"),
-		_("2.5x Native (1600x1320)"), _("3x Native (1920x1584)"), _("4x Native (2560x2112)"), _("Custom") };
+		_("Native (640x528)"), _("1.5x Native (960x792)"), _("2x Native (1280x1056) for 720p"), _("2.5x Native (1600x1320)"),
+		_("3x Native (1920x1584) for 1080p"), _("4x Native (2560x2112) for 1440p"), _("5x Native (3200x2640)"),
+		_("6x Native (3840x3168) for 4K"), _("7x Native (4480x3696)"), _("8x Native (5120x4224) for 5K"), _("Custom") };
```
Comment (degasus) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: For translation, I'd prefer _("%dx Native (%dx%d)")

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 4]
```diff
@@ -14,6 +14,7 @@
 
 #include <cinttypes>
 #include <cmath>
+#include <sstream>
```
Comment (lioncash) [inline Source/Core/VideoCommon/RenderBase.cpp]: this include isn't needed.

**Comment by Kodiack** [comment]
@karasuhebi, the reason that it's sharper is because of Wind Waker's "depth of field" effect. Said depth of field effect can be fully disabled with a projection hack via an INI setting, which will heavily negate the benefits of increasing IR beyond your monitor's native resolution. Anyone looking for better anti-aliasing would be much better served using SSAA if their IR already matches their monitor's resolution. SSAA does actually filtering while downsampling, whereas increasing IR is a bit more naive. You'll get a "sharper" image, but you're still prone to realise aliasing.

Here's a link to some projection hack INI settings that will remove the DOF in Wind Waker; give 'em a shot and you'll see that extra IR isn't nearly as beneficial as it may first appear:

https://forums.dolphin-emu.org/Thread-gc-the-legend-of-zelda-the-wind-waker--25929?pid=364125#pid364125

I also had a patch over a year ago that allowed up to 6xIR. I rather liked neobrain's stance with that.

https://forums.dolphin-emu.org/Thread-patch-5x-3200x2640-and-6x-3840x3168-native-ir

Things really are starting to look a bit ridiculous. I think that a slider for IR may be a good idea. There could be an exponential slider, much like what Fiora implemented for overclocking the emulated CPU. Perhaps it would even be possible to add in a text box next to it, where people could punch in a specific IR and have the slider snap to that value.

**Comment by Icekhaos** [comment]
+1 to @Kodiack 's suggestion for a slider and box, but to also have a vertical box.

**Pull request merged**